### PR TITLE
lease: forward lease keepalive to leader

### PR DESCRIPTION
**Pull request by heyitsanthony**
depends on PR #4253

### CODE REVIEW

File: etcdserver/etcdhttp/peer.go [hunk 1]
```diff
@@ -19,15 +19,21 @@ import (
...
 )
 
-// NewPeerHandler generates an http.Handler to handle etcd peer (raft) requests.
-func NewPeerHandler(cluster etcdserver.Cluster, raftHandler http.Handler) http.Handler {
+// NewPeerHandler generates an http.Handler to handle etcd peer requests.
+func NewPeerHandler(s *etcdserver.EtcdServer) http.Handler {
+	return newPeerHandler(s.Cluster(), s.RaftHandler(), s.Lessor())
+}
+
+func newPeerHandler(cluster etcdserver.Cluster, raftHandler http.Handler, lessor lease.Lessor) http.Handler {
```
Comment (xiang90) [inline etcdserver/etcdhttp/peer.go]: can we change lessor type arg to lessorHandler for this handler func? And then create the lessor Handler in `NewPeerHandler` func at line 33?

File: lease/http.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+// RenewHTTP renews a lease at a given primary server.
+func RenewHTTP(id LeaseID, url string, tlsInfo transport.TLSInfo, timeout time.Duration) (int64, error) {
+	// will post lreq protobuf to leader
+	lreq, err := (&pb.LeaseKeepAliveRequest{ID: int64(id)}).Marshal()
+	if err != nil {
+		return -1, err
+	}
+
+	rt, err := transport.NewTimeoutTransport(tlsInfo, timeout, 0, 0)
```
Comment (xiang90) [inline lease/http.go]: creating a new transporter for each forwarded request seems to be very expensive. 

We probably should add a todo for fixing this. Also we probably also want to batch renew forwarding requests in the future.

File: lease/http.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+		// TODO detect if leader failed and retry?
+		return -1, err
+	}
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return -1, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return -1, fmt.Errorf("%s", string(b))
```
Comment (xiang90) [inline lease/http.go]: add lease prefix for error.

File: lease/http.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	}
+
+	lresp := &pb.LeaseKeepAliveResponse{}
+	if err := lresp.Unmarshal(b); err != nil {
+		return -1, fmt.Errorf(`%v. data = "%s"`, err, string(b))
```
Comment (xiang90) [inline lease/http.go]: same here.

File: lease/http.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+	}
+	if lresp.ID != int64(id) {
+		return -1, fmt.Errorf("id mismatch")
```
Comment (xiang90) [inline lease/http.go]: same here.and probably change to renew ID mismatch. (which should never happen though)

**Comment by heyitsanthony** [comment]
ok all fixed up /cc @xiang90

**Comment by xiang90** [comment]
LGTM after fixing tests.

**Comment by xiang90** [comment]
Also please squash.

**Pull request merged**